sys/include/net/sock/dtls.h

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright (C) 2019 HAW Hamburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    net_sock_dtls    DTLS sock API
+ * @ingroup     net_sock
+ * @brief       Sock submodule for DTLS
+ * @{
+ *
+ * @file
+ * @brief   DTLS sock definitions
+ *
+ * @author  Aiman Ismail <muhammadaimanbin.ismail@haw-hamburg.de>
+ */
+
+#ifndef NET_SOCK_DTLS_H
+#define NET_SOCK_DTLS_H
+
+#include <assert.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <sys/types.h>
+
+#include "net/sock.h"
+#include "net/sock/udp.h"
+#include "net/credman.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief   Type for a DTLS sock object
+ *
+ * @note    API implementors: `struct sock_dtls` needs to be defined by
+ *          implementation-specific `sock_dtls_types.h`.
+ */
+typedef struct sock_dtls sock_dtls_t;
+
+/**
+ * @brief Information about the established session with
+ *        the remote endpoint. Used when sending and
+ *        receiving data to the endpoint
+ */
+typedef struct sock_dtls_session sock_dtls_session_t;
+
+/**
+ * @brief Must be called once before any other use
+ */
+void sock_dtls_init(void);
+
+/**
+ * @brief Creates a new DTLS sock object
+ *
+ * @param[out] sock     The resulting DTLS sock object
+ * @param[in] udp_sock  Existing UDP sock to be used
+ * @param[in] tag       Credential tag of the sock. Used to get the right
+ *                      credential from pool.
+ * @param[in] method    Defines the method for the client or server to use.
+ *
+ * @return  0 on success.
+ * @return  value < 0 on error
+ */
+int sock_dtls_create(sock_dtls_t *sock, sock_udp_t *udp_sock,
+                     credman_tag_t tag,unsigned method);
+
+/**
+ * @brief Initialises the server to listen for incoming connections
+ *
+ * @param[in] sock      DTLS sock to listen to
+ */
+void sock_dtls_init_server(sock_dtls_t *sock);
+
+/**
+ * @brief Establish DTLS session to a server. Execute the handshake step in
+ *        DTLS.
+ *
+ * @param[in]  sock     DLTS sock to use
+ * @param[in]  ep       Endpoint to establish session with
+ * @param[out] remote   The established session, cannot be NULL
+ *
+ * @return  0 on success
+ * @return  value < 0 on other errors
+ * @return  -EADDRNOTAVAIL, if sock_udp_ep_t::local of @p sock->udp_sock is not
```

Comment:
`sock_udp_ep_t` does not have a member `local`

Reply:
This supposed to be sock_udp_t. Fixed.

Reply:
Then this is wrong. Because `sock_udp_t` is not guaranteed to have a member `local` either. It has for the GNRC and lwIP implementations, but it is not guaranteed other implementations will have that. Come to think of it `sock->udp_sock` is _also_ implementation specific (in this case to your `tinydtls` implementation. I'd rather either add the local endpoint as an argument or word it more generic, e.g. "[…], if the local endpoint of the UDP sock of @p sock is was not set".

Reply:
> [...] word it more generic, e.g. "[…], if the local endpoint of the UDP sock of @p sock is was not set".

I think this will do. This way there is no need to add to the parameter list, which is quite long already.